org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/ServiceUseRuleTest.java

```diff
@@ -0,0 +1,317 @@
+				.isEqualTo(ServiceUse.DEFAULT_TIMEOUT);
+			softly.assertThat(foos.getCardinality())
+				.isEqualTo(0);
+
+			softly.assertAll();
+		}
+	}
+
+	@Test
+	public void requiredFailsWhenNoService() throws Exception {
+		Throwable throwable = catchThrowable(() -> {
+			try (BundleContextRule bcRule = new BundleContextRule();
+				ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).cardinality(1)
+					.build()) {
+
+				bcRule.init(getClass());
+				foos.init(getClass());
+			}
+		});
+
+		assertThat(throwable).isInstanceOf(AssertionError.class)
+			.hasMessageContaining(
+				" services (objectClass=org.osgi.test.junit4.types.Foo) didn't arrive within 200ms");
+	}
+
+	@Test
+	public void requiredFailsWhenNoServiceWithTimeout() throws Exception {
+		Throwable throwable = catchThrowable(() -> {
+			try (BundleContextRule bcRule = new BundleContextRule();
+				ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).timeout(10)
+					.build()) {
+
+				bcRule.init(getClass());
+				foos.init(getClass());
+			}
+		});
+
+		assertThat(throwable).isInstanceOf(AssertionError.class)
+			.hasMessageContaining(
+				" services (objectClass=org.osgi.test.junit4.types.Foo) didn't arrive within 10ms");
+	}
+
+	@Test
+	public void successWhenService() throws Exception {
+		try (BundleContextRule bcRule = new BundleContextRule();
+			ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).cardinality(1)
+				.build();) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					100, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			SoftAssertions softly = new SoftAssertions();
+
+			softly.assertThat(foos.isEmpty())
+				.isFalse();
+			softly.assertThat(foos.size())
+				.isEqualTo(1);
+			softly.assertThat(foos.getTimeout())
+				.isEqualTo(ServiceUse.DEFAULT_TIMEOUT);
+			softly.assertThat(foos.getCardinality())
+				.isGreaterThan(0);
+
+			softly.assertAll();
+		}
+	}
+
+	@Test
+	public void successWhenServiceWithTimeout() throws Exception {
+		try (BundleContextRule bcRule = new BundleContextRule();
+			ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule)
+				.timeout(10)
+				.build()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					5, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			SoftAssertions softly = new SoftAssertions();
+
+			softly.assertThat(foos.getService())
+				.isNotNull();
+			softly.assertThat(foos.getServiceReference())
+				.isNotNull();
+			softly.assertThat(foos.getServiceReferences())
+				.isNotNull();
+			softly.assertThat(foos.getServices())
+				.isNotEmpty()
+				.contains(afoo);
```

Comment:
`isNotEmpty()` is redundant if you're asserting `contains(afoo)`.

Reply:
fixing...